src/transaction.cpp

```diff
@@ -3,52 +3,62 @@
 
 #include <rlpvalue.h>
 #include <iostream>
+#include <exception>
 
 /**
-* Returns the raw Bytes of the EIP-1559 transaction, in order.
-*
-* Format: `[chainId, nonce, maxPriorityFeePerGas, maxFeePerGas, gasLimit, to, value, data,
-* accessList, signatureYParity, signatureR, signatureS]`
-*
-* For an unsigned tx this method uses the empty Bytes values for the
-* signature parameters `v`, `r` and `s` for encoding.
-*/
+ * Serialize the transaction into a hexadecimal string using RLP encoding.
+ * This function handles both signed and unsigned transactions.
+ *
+ * Throws: std::runtime_error on failure to process any transaction component. 
+ * Returns the raw Bytes of the EIP-1559 transaction, in order.
+ *
+ * Format: `[chainId, nonce, maxPriorityFeePerGas, maxFeePerGas, gasLimit, to, value, data,
+ * accessList, signatureYParity, signatureR, signatureS]`
+ *
+ * For an unsigned tx this method uses the empty Bytes values for the
+ * signature parameters `v`, `r` and `s` for encoding.
+ */
 std::string Transaction::serialized() const {
-    RLPValue arr(RLPValue::VARR);
-    arr.setArray();
-    RLPValue temp_val;
-    temp_val.clear();
-    temp_val.assign(utils::intToBytes(chainId));
-    arr.push_back(temp_val);
-    temp_val.assign(utils::intToBytes(nonce));
-    arr.push_back(temp_val);
-    temp_val.assign(utils::intToBytes(maxPriorityFeePerGas));
-    arr.push_back(temp_val);
-    temp_val.assign(utils::intToBytes(maxFeePerGas));
-    arr.push_back(temp_val);
-    temp_val.assign(utils::intToBytes(gasLimit));
-    arr.push_back(temp_val);
-    temp_val.assign(utils::fromHexString(to));
-    arr.push_back(temp_val);
-    temp_val.assign(utils::intToBytes(value));
-    arr.push_back(temp_val);
-    temp_val.assign(utils::fromHexString(data));
-    arr.push_back(temp_val);
+    try {
+        RLPValue arr(RLPValue::VARR);
+        arr.setArray();
 
-    // Access list not going to use
-    RLPValue access_list(RLPValue::VARR);
-    access_list.setArray();
-    arr.push_back(access_list);
+        // Helper lambda for repetitive operations to convert and append data
+        auto appendSerializedData = [&](const auto& data, auto conversionFunction) {
```

Comment:
the data parameter here shadows the `Transaction.data` member. `-Werror` is set so this is causing the tests to fail atm

Reply:
Sorry about that. I don't know why I didn't catch that. 
I pushed a fix for that up to this branch and made a few more optimizations. 
(compiler hint to inline the helper function, pre-allocation of memory, and some efficient conditional logic.)

Thinking about it now... Have you considered explicitly setting all transaction-related defaults in the constructor? This would ensure every transaction starts with a known good state. 

```
Transaction(const std::string& _to, uint64_t _value, uint64_t _gasLimit = 21000, const std::string& _data = "", uint64_t _chainId = 1, uint64_t _nonce = 0, uint64_t _maxPriorityFeePerGas = 1, uint64_t _maxFeePerGas = 1)
    : chainId(_chainId), nonce(_nonce), maxPriorityFeePerGas(_maxPriorityFeePerGas), maxFeePerGas(_maxFeePerGas), to(_to), value(_value), gasLimit(_gasLimit), data(_data) {}

```

Reply:
The goal with the constructor was just to require the bare minimum needed to get a transaction going. I was assuming that most of the other items were going to be set by the providers defaults (ie id say its pretty rare that the user wants to manually set chainID but instead want to rely on the node to provide it), so putting them into a known bad state means that we can check if it has changed before we query the node.

I do like that the constructor at the moment is minimal, but it does assume that the default pathway will be for it to go through the signer/provider and things will be changed. 
```
    Transaction tx("0x2Ccb8b65024E4aA9615a8E704DFb11BE76674f1F", 100000000000000);
    const auto hash = signer.sendTransaction(tx, seckey);
```

However I do see the point, tx above does not immediately come into a ready state. It would need extra steps like this
```
tx.chainId = 1;
tx.nonce = 1;
```
If you for example wanted the tx to be exported immediately without the signer/provider cleaning things up.